Fix #21160 having geojson in protocol and file twice (add geojsonseq)

rduivenvoorde opened the pull request:
## Description

OGR provides a driver list with geojson and geojsonseq
Going over the list we only tested if the name 'startsWith',
resulting in having geojson double.

Split up to work with both geojson (*.geojson) and
geojsonseq (*.geojsonl and *.geojsons)

See: https://www.gdal.org/drv_geojsonseq.html in short: instead of a FeatureCollection with Features, 
you stream one type (probably only Features) sequentially separated with newlines. Example: https://duif.net/mw.geojsonl

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [ X] Commit messages are descriptive and explain the rationale for changes
- [ X] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ X] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [ X] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [ X] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

rduivenvoorde committed:
Fix #21160: fix geojson in protocol and file twice

GDAL/OGR has two geojson related drivers now: GeoJSON and
GeoJSONSeq, see https://www.gdal.org/drv_geojsonseq.html

This commit makes it possible to open GeoJSONSeq (geojsonl and
geojsons and json) files as protocols/streams. And write them
as *.json.

rduivenvoorde commented:
If you want to test, there are 3 small geojson files available of a car driving on a highway in NL:
https://duif.net/mw.geojson https://duif.net/mw.geojsons https://duif.net/mw.geojsonl

Questions: I've been looking over this code to other parts which mention 'geojson': 

https://github.com/qgis/QGIS/blob/master/src/providers/ogr/qgsogrprovider.cpp#L536
I think this is not necessary as geojsonseq driver was introduced in GDAL 2.4 see:
https://www.gdal.org/drv_geojsonseq.html

https://github.com/qgis/QGIS/blob/master/src/providers/ogr/qgsogrprovider.cpp#L4466
For that one I'm not sure, please let me know because if this is hit (yet) because I'm not sure 
what the mDynamicWriteAccess is needed for...
But happy to change that to `mGDALDriverName.startsWith( QLatin1String( "GeoJSON" ) )`
too. OR is that only needed when you WRITE a geojson[sl] file?

This is a bugfix, but if I also have a look into the writing of it maybe it is a feature?

rduivenvoorde commented:
This commit adds the necessary metadata to see GeoJSONSeq in the export As dialogs.
Note that according to https://www.gdal.org/drv_geojsonseq.html there is a slight difference between writing a *.geojsonl and a *.geojsons file. The First one is the default ( RS=NO ). The second one could be written if RS=YES and should result in a file with the *.geosjons this adds a newline IN FRONT of every feature (instead of on the end).
I'm not sure if there is a desire to be able to do this (@rouault what do you think?). 
IF it is desired ten I wonder if this should be done by creating another metadata block as I did now, OR if this could be done within current added metadata block. Note that we should be able to define 2 file-extensions then (.geojson. and .geojsons) 
AND depending on the value of RS then decide which one to use
AND depending on the file name then set the right RS value.
So I though to stick with the default for GDAL now. Maybe I should add some comments about this in the source?

rouault commented:
> with the *.geosjons this adds a newline IN FRONT of every feature (instead of on the end).

Not quite. This adds a RS character in front of every feature. And a LF at the end.

Honestly I'm not sure if it is worth cluttering QGIS UI with such a niche format in creation (my understanding is that is was more aimed at exchanging data between network services), and its two subvariants, but I don't have a strong opinion either. Perhaps if we expose RS=NO this should be called "Newline delimited JSON" since GeoJSONSeq might make thing to GeoJSON Text Sequences (RS separator)

rduivenvoorde commented:
Then best to leave it like it is now: the 'geojsonseq' driver creates a *.geojsonl file, which I think is a nice option to have: 1) there is no double 'geojson'-option and 2) sometimes you wonder what such a format looks like (or you just need it for something else, as I did).
So if people really want to have a geojsons, they can add RS=YES as custom option I think and rename there file, because we always add prepend the filename with .geojsonl if it is something else that.

nyalldawson commented:
So whats the status here? Should this be closed?

rduivenvoorde commented:
@nyalldawson No please, do not close it, either just pull or I fix what people want.

Current PR fixes at least the twice available Geojson's in the type dropdowns when you want to open or expart a geojson file either as file or as protocol:

![image](https://user-images.githubusercontent.com/731673/53081890-0615b680-34fc-11e9-9ddb-d4e434f3e048.png)

Is now for opening:

![image](https://user-images.githubusercontent.com/731673/53087665-fac88800-3507-11e9-92a9-ea254ea5e983.png)

Hmmm, I see there is a double Esri Shapefiles there too for opening ??
And to save: 

![gnome-shell-screenshot-0lbgxz](https://user-images.githubusercontent.com/731673/53088019-de791b00-3508-11e9-9a08-c8b644ce6230.png)

Standard will make it possible to Open/Save a *.geojson or a *.geojsonL file. What it does NOT do is to make it easy to SAVE a *.geojsonS file (gdal will OPEN all geojso* files fine anyway)

If @rouault meant that adding *.geojsonl and/or *.geojsons is cluttering, I do not agree. I think it can be handy to be able to save at least one of the GeoJSONSeq versions (I personally had this urge already since I know of the dependence of this format :-) )  I choose to only add GeoJSONSeq (as geojsonl)

So Questions are:
- is this PR good enough as is => just pull
- does it need polishing (use "Newline delimited JSON"  instead of GeoJSONSeq)?
- do I need to refactor it so there will be a 3 geojson-items in the dropdowns

DelazJ commented:
> Hmmm, I see there is a double Esri Shapefiles there too for opening ??

See https://issues.qgis.org/issues/20127 (_just in case_)

luipir commented:
> * is this PR good enough as is => just pull

IMHO it's ok (other than adding geojsonl)

> * does it need polishing (use "Newline delimited JSON"  instead of GeoJSONSeq)?

I would prefer

> * do I need to refactor it so there will be a 3 geojson-items in the dropdowns

IMHO yes

my2c

rduivenvoorde commented:
@DelazJ the double Esri Shapefile is like this issue: besides the "Esri" driver there is also an "ESRIJSON" driver (both starting with ESRI). We could split that up like I did here, OR we can hide the esrijson one.
It all depends if we as QGIS want to be 'feature complete' with GDAL, or that we should skip some exotic drivers to not clutter the QGIS Gui (as @rouault says). 

Reason I split up the geojson - geojsonseq  combi is because I think I can use it for huge datasets  I have.

rduivenvoorde commented:
@luipir @nyalldawson ok I changed the commits (and squashed to one) to give user more flexibility without adding too many options:

* The vector FILE open dialog now shows 3 'geojson' options, because there it is important which filter to choose. Eg the Newline Delimited JSON has filter *.geojsonl while the other has *.geojsons:
I added the [GeoJSONSeq] so it is clear which driver is used.
I added Geojson in front to group them together (as the list is sorted)

![gnome-shell-screenshot-9thvxz](https://user-images.githubusercontent.com/731673/53266771-0ff90e80-36e2-11e9-9346-801343c16fda.png)

* The vector PROTOCOL open dialog shows 2 'geojson' options because whatever url you provide all are opened by the GeoJSONSeq driver (so no need to differentiate between the two subvariants):

![gnome-shell-screenshot-badxxz](https://user-images.githubusercontent.com/731673/53266964-8b5ac000-36e2-11e9-9403-ba910a7c9cef.png)

* The Export Features dialog shows 2 'geojson' options, where the second one (selected in img below) has an extra Layer Option (RS=NO/YES) to determine if you save it without RS=0x1E character (*.geojsonl) or with (*.geosjons) RS=0x1E character on the start of the record.

![gnome-shell-screenshot-pyjtxz](https://user-images.githubusercontent.com/731673/53267077-cc52d480-36e2-11e9-8911-8c28c82e98a5.png)

I hope this is clear enough?

nyalldawson commented on src/core/qgsvectorfilewriter.cpp:
```diff
@@ -1022,6 +1022,35 @@ class QgsVectorFileWriterMetadataContainer
+
+      layerOptions.insert( QStringLiteral( "COORDINATE_PRECISION" ), new QgsVectorFileWriter::IntOption(
+                             QObject::tr( "Maximum number of figures after decimal separator to write in coordinates. "
+                                          "Default to 15. Truncation will occur to remove trailing zeros." ),
```
```suggestion
                                          "Defaults to 15. Truncation will occur to remove trailing zeros." ),
```

nyalldawson commented:
I'd suggest instead of "GeoJSON Newline Delimited JSON [GeoJSONSEQ]" going with "GeoJSON - Newline Delimited". It removes the duplicate "JSON" text, and I don't think we need to expose the internal GDAL driver name here, which has little context for the majority of users.

rduivenvoorde commented:
@nyalldawson thanks. I'll do that (and also fix the spelling mistake which makes Travis fail). On hindsight I think it is even better in the Open dialog, to NOT split it up in the 2 versions either, so all looks the same:   "GeoJSON - Newline Delimited"

One thing what bothers me is the issue of file extensions: because I do not know if a user picks RS=YES or RS=NO I do not know which extension to use (geojsonl or geojsons or even other options: nlgeojon (newline geojson). So i ended up with NOT using a file extension... 
Any idea for that? Could be nice if you could add several extensions in the Metadata with the first one preferred so we could check if the filename ends on one of the extensions and if not: add the prefered one. 
Another thing that bothers me (on Linux) is that it does NOT show the *globs in the file open dialogs so there is no way to see wat you are looking for. While in contrast Window shows (*.shp,*.SHP) (which we add ourselves) while windows is not even caseaware (it will open *.ShP also anyway.
Will ask this on dev list maybe.

rduivenvoorde commented:
Bumped Travis. Now it looks like this:if you want to open a file:

![gnome-shell-screenshot-iz7tyz](https://user-images.githubusercontent.com/731673/53979813-05da0580-410f-11e9-9cb7-e139cb70b333.png)

You can open all kind of geojson files

To open a protocol looks the same:


![gnome-shell-screenshot-6lfcyz](https://user-images.githubusercontent.com/731673/53979867-26a25b00-410f-11e9-87a8-e96200b4a0cb.png)

To save you see: and you can choose to add a RS character or NOT:

![gnome-shell-screenshot-sgjuyz](https://user-images.githubusercontent.com/731673/53979958-69643300-410f-11e9-9096-f418ee5f5ae4.png)

nyalldawson commented:
Thanks @rduivenvoorde !
